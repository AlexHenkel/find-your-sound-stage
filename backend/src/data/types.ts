export interface StudioData {
  id: number;
  studio_name: string;
  city: string;
  state: string;
  country: string;
  sound_stages: SoundStage[];
}

interface SoundStage {
  stage_id: number;
  stage_number: string;
  available_dates: AvailableDate[];
}

interface AvailableDate {
  id: number;
  start_date: string;
  end_date: string | null;
}
