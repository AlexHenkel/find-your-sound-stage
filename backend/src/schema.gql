# -----------------------------------------------
# !!! THIS FILE WAS GENERATED BY TYPE-GRAPHQL !!!
# !!!   DO NOT MODIFY THIS FILE BY YOURSELF   !!!
# -----------------------------------------------

"""
The javascript `Date` as string. Type represents date and time as the ISO Date string.
"""
scalar DateTime

type Mutation {
  search(query: SearchInput): [Studio!]!
}

type Query {
  healthCheck: Boolean!
}

input SearchInput {
  city: String
  country: String
  endDate: DateTime
  startDate: DateTime
  state: String
}

type SoundStage {
  stageId: Int!
  stageNumber: String!
}

type Studio {
  availableStage: SoundStage!
  city: String!
  country: String!
  id: Int!
  state: String!
  studioName: String!
}
