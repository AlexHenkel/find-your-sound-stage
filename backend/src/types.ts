import { Field, ObjectType, Int } from "type-graphql";

@ObjectType()
class SoundStage {
  @Field(() => Int)
  stageId: number;

  @Field()
  stageNumber: string;
}

@ObjectType()
export class Studio {
  @Field(() => Int)
  id: number;

  @Field()
  studioName: string;

  @Field()
  city: string;

  @Field()
  state: string;

  @Field()
  country: string;

  @Field(() => SoundStage)
  availableStage: SoundStage;
}
