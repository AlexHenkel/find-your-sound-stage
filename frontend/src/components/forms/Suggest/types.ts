export interface Option {
  label: string;
  value: string;
}

/**
 * If this component requires further customization, consider extending original props
 * to allow overriding every original prop
 */
export interface SuggestProps {
  label: string;
  options: Option[];
  onSelect: (value: string) => void;
  value?: string;
}
