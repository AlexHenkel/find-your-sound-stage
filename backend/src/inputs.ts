import { InputType, Field } from "type-graphql";

@InputType()
export class SearchInput {
  @Field({ nullable: true })
  city?: string;

  @Field({ nullable: true })
  state?: string;

  @Field({ nullable: true })
  country?: string;

  @Field({ nullable: true })
  startDate?: Date;

  @Field({ nullable: true })
  endDate?: Date;
}
