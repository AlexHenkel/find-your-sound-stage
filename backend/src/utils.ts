import { isAfter, isBefore, startOfDay } from "date-fns";
import { Studio } from "./types";
import { SearchInput } from "./inputs";
import { StudioData } from "./data/types";

type SearchString = keyof SearchInput & keyof Studio;
const stringKeys: SearchString[] = ["city", "state", "country"];

export const filterStages = (
  searchQuery: SearchInput | undefined,
  data: StudioData[]
) =>
  searchQuery
    ? data.filter((studio) =>
        // Either filter not present, or filter match with value
        stringKeys.every(
          (k) =>
            !searchQuery[k] ||
            // Use lower case to expand results
            studio[k].toLowerCase() === searchQuery[k]?.toLowerCase()
        )
      )
    : // If query not present, just return all data
      data;

export const filterDates = (
  searchQuery: SearchInput | undefined,
  filteredStudios: StudioData[]
) => {
  // Make sure we compare with the start of day, so `isBefore` or `isAfter` are false with same dates
  const queryStartDate = searchQuery?.startDate
    ? startOfDay(searchQuery.startDate)
    : undefined;
  const queryEndDate = searchQuery?.endDate
    ? startOfDay(searchQuery.endDate)
    : undefined;

  const response: Studio[] = [];

  // Search for matching dates on filtered studios
  filteredStudios.forEach((studio) => {
    studio.sound_stages.forEach((stage) => {
      // Use regular for, so we can break and stop searching for dates on same stage
      for (const date of stage.available_dates) {
        // Either filter not present, or filter date is same or after start date
        if (
          !queryStartDate ||
          !isBefore(queryStartDate, startOfDay(new Date(date.start_date)))
        ) {
          // Either filter not present, or infinite availability, or filter date is same or before end date
          if (
            !queryEndDate ||
            !date.end_date ||
            !isAfter(queryEndDate, startOfDay(new Date(date.end_date)))
          ) {
            // Construct response and stop looking for available dates
            response.push({
              id: studio.id,
              city: studio.city,
              state: studio.state,
              country: studio.country,
              studioName: studio.studio_name,
              availableStage: {
                stageId: stage.stage_id,
                stageNumber: stage.stage_number,
              },
            });
            break;
          }
        }
      }
    });
  });

  return response;
};
