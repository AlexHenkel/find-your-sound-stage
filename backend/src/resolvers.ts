import { Resolver, Arg, Mutation, Query } from "type-graphql";
import { Studio } from "./types";
import { SearchInput } from "./inputs";
import { filterDates, filterStages } from "./utils";
import data from "./data/stageData.json";

@Resolver(() => Studio)
export class StudioResolver {
  @Mutation(() => [Studio])
  async search(
    @Arg("query", { nullable: true })
    searchQuery?: SearchInput
  ): Promise<Studio[]> {
    // First get available studios based on stage filters, then filter dates per stage
    return filterDates(searchQuery, filterStages(searchQuery, data));
  }

  @Query(() => Boolean)
  healthCheck(): boolean {
    return true;
  }
}
