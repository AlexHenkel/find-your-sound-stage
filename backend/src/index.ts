import "reflect-metadata";
import * as path from "path";
import { ApolloServer } from "apollo-server";
import { buildSchema } from "type-graphql";
import { StudioResolver } from "./resolvers";

async function bootstrap() {
  const schema = await buildSchema({
    resolvers: [StudioResolver],
    emitSchemaFile: path.resolve(__dirname, "schema.gql"),
  });

  const server = new ApolloServer({
    schema,
  });

  const { url } = await server.listen(4000);

  // eslint-disable-next-line no-console
  console.log(`Server is running, GraphQL available at ${url}`);
}

bootstrap();
