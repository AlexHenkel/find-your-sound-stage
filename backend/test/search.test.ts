import { filterDates, filterStages } from "../src/utils";
import { SearchInput } from "../src/inputs";
import { StudioData } from "../src/data/types";
import data from "../src/data/stageData.json";
import { startOfDay } from "date-fns";

test("Filter stages correctly with city filter", () => {
  const testSearch: SearchInput = { city: "Los Angeles" };
  const res = filterStages(testSearch, data);

  expect(res).toMatchSnapshot();
  expect(res.every(({ city }) => city === "Los Angeles")).toBe(true);
});

test("Filter stages correctly with city and state filters", () => {
  const testSearch: SearchInput = { city: "Los Angeles", state: "California" };
  const res = filterStages(testSearch, data);

  expect(res).toMatchSnapshot();
  expect(
    res.every(
      ({ city, state }) => city === "Los Angeles" && state == "California"
    )
  ).toBe(true);
});

test("Filter stages correctly with city, state and country filters", () => {
  const testSearch: SearchInput = {
    city: "Los Angeles",
    state: "California",
    country: "United States of America",
  };
  const res = filterStages(testSearch, data);

  expect(res).toMatchSnapshot();
  expect(
    res.every(
      ({ city, state, country }) =>
        city === "Los Angeles" &&
        state == "California" &&
        country === "United States of America"
    )
  ).toBe(true);
});

test("Empty results with wrong filter", () => {
  const testSearch: SearchInput = {
    city: "Legoland",
    state: "California",
    country: "United States of America",
  };
  const res = filterStages(testSearch, data);

  expect(res).toMatchSnapshot();
  expect(res).toHaveLength(0);
});

const mockData: StudioData[] = [
  {
    id: 1,
    studio_name: "Sunset Bronson",
    city: "Los Angeles",
    state: "California",
    country: "United States of America",
    sound_stages: [
      {
        stage_id: 1,
        stage_number: "1A",
        available_dates: [
          {
            id: 111,
            start_date: "2020-01-20",
            end_date: "2020-03-15",
          },
          {
            id: 112,
            start_date: "2020-05-01",
            end_date: null,
          },
        ],
      },
    ],
  },
];

test("Get all dates with no date filters", () => {
  const res = filterDates({}, filterStages({}, mockData));
  expect(res).toHaveLength(1);
});

test("Get dates with start and end filter", () => {
  const testSearch: SearchInput = {
    startDate: startOfDay(new Date("2020-01-28")),
    endDate: startOfDay(new Date("2020-03-01")),
  };
  const res = filterDates(testSearch, filterStages(testSearch, mockData));
  expect(res).toHaveLength(1);
});

test("Get no dates with start filter out of range", () => {
  const testSearch: SearchInput = {
    startDate: startOfDay(new Date("2020-01-19")),
    endDate: startOfDay(new Date("2020-03-01")),
  };
  const res = filterDates(testSearch, filterStages(testSearch, mockData));
  expect(res).toHaveLength(0);
});

test("Get no dates with end filter out of range", () => {
  const testSearch: SearchInput = {
    startDate: startOfDay(new Date("2020-01-28")),
    endDate: startOfDay(new Date("2020-03-16")),
  };
  const res = filterDates(testSearch, filterStages(testSearch, mockData));
  expect(res).toHaveLength(0);
});

test("Get dates with start and infinite end", () => {
  const testSearch: SearchInput = {
    startDate: startOfDay(new Date("2020-05-02")),
    endDate: startOfDay(new Date("2023-05-03")),
  };
  const res = filterDates(testSearch, filterStages(testSearch, mockData));
  expect(res).toHaveLength(1);
});
