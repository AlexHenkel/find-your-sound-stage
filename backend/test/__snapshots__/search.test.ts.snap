// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Empty results with wrong filter 1`] = `Array []`;

exports[`Filter stages correctly with city and state filters 1`] = `
Array [
  Object {
    "city": "Los Angeles",
    "country": "United States of America",
    "id": 1,
    "sound_stages": Array [
      Object {
        "available_dates": Array [
          Object {
            "end_date": "2020-03-15",
            "id": 111,
            "start_date": "2020-01-20",
          },
          Object {
            "end_date": null,
            "id": 112,
            "start_date": "2020-05-01",
          },
        ],
        "stage_id": 1,
        "stage_number": "1A",
      },
      Object {
        "available_dates": Array [
          Object {
            "end_date": "2020-07-10",
            "id": 113,
            "start_date": "2020-02-08",
          },
          Object {
            "end_date": null,
            "id": 114,
            "start_date": "2021-01-01",
          },
        ],
        "stage_id": 2,
        "stage_number": "1B",
      },
    ],
    "state": "California",
    "studio_name": "Sunset Bronson",
  },
]
`;

exports[`Filter stages correctly with city filter 1`] = `
Array [
  Object {
    "city": "Los Angeles",
    "country": "United States of America",
    "id": 1,
    "sound_stages": Array [
      Object {
        "available_dates": Array [
          Object {
            "end_date": "2020-03-15",
            "id": 111,
            "start_date": "2020-01-20",
          },
          Object {
            "end_date": null,
            "id": 112,
            "start_date": "2020-05-01",
          },
        ],
        "stage_id": 1,
        "stage_number": "1A",
      },
      Object {
        "available_dates": Array [
          Object {
            "end_date": "2020-07-10",
            "id": 113,
            "start_date": "2020-02-08",
          },
          Object {
            "end_date": null,
            "id": 114,
            "start_date": "2021-01-01",
          },
        ],
        "stage_id": 2,
        "stage_number": "1B",
      },
    ],
    "state": "California",
    "studio_name": "Sunset Bronson",
  },
]
`;

exports[`Filter stages correctly with city, state and country filters 1`] = `
Array [
  Object {
    "city": "Los Angeles",
    "country": "United States of America",
    "id": 1,
    "sound_stages": Array [
      Object {
        "available_dates": Array [
          Object {
            "end_date": "2020-03-15",
            "id": 111,
            "start_date": "2020-01-20",
          },
          Object {
            "end_date": null,
            "id": 112,
            "start_date": "2020-05-01",
          },
        ],
        "stage_id": 1,
        "stage_number": "1A",
      },
      Object {
        "available_dates": Array [
          Object {
            "end_date": "2020-07-10",
            "id": 113,
            "start_date": "2020-02-08",
          },
          Object {
            "end_date": null,
            "id": 114,
            "start_date": "2021-01-01",
          },
        ],
        "stage_id": 2,
        "stage_number": "1B",
      },
    ],
    "state": "California",
    "studio_name": "Sunset Bronson",
  },
]
`;
